mac80211: fix missing defines

oskarirauta opened the pull request:
fixes build issue with missing defines (archlinux 2022)

my build failed due to missing lock state defines, this patch fixes the problem.

Signed-off-by: Oskari Rauta <oskari.rauta@gmail.com>

oskarirauta committed:
mac80211: fix missing defines

fixes failure during build issue due to missing
lock state definitions.

Signed-off-by: Oskari Rauta <oskari.rauta@gmail.com>

ynezz commented:
> fixes build issue with missing defines (archlinux 2022)

I fail to see how this is related to the host distro, you should add relevant build failure in the commit message.

oskarirauta commented:
Maybe not relevant to host, but it is mentioned for possible replication of issue.

oskarirauta commented:
commit message adjusted. I am not going to remove my patch and re-compile whole kernel just to show a log of failing build with unknown definitions/variables - we all know what that looks like... Not sure why this happens as this isn't brand new issue, I've had this in past as well - but if you check the patch, you see that it doesn't hurt anyone as if these are already defined in some build scenario, they are ignored because of ifndef directive. Proper values were derived from kernel source code.

oskarirauta commented:
@ynezz ping

ynezz commented:
> we all know what that looks like...

No I don't for example, otherwise I wouldn't ask you for that. It shouldn't fail, otherwise I assume, that it would fail consistently everywhere. So I assume, that there is some host leakage happening somewhere and it should be probably visible in the compiler invocation/error message.

oskarirauta commented:
> > we all know what that looks like...
> 
> No I don't for example, otherwise I wouldn't ask you for that. It shouldn't fail, otherwise I assume, that it would fail consistently everywhere. So I assume, that there is some host leakage happening somewhere and it should be probably visible in the compiler invocation/error message.

Well, it looks somewhat like this:
`error: 'FOO' was not declared in this scope`

Listing source code in backports reveals that these defines are nowhere to be found, so sure, there's error; how compiler know what _FOO_ is if it's not properly introduced.

> In other words, your proposed fix is some kind of workaround, not a proper fix.

If you look at that linked kernel patch, you know it is a actual fix, as it's in kernel defined with same values - it is not my hand writing, I just mimicked missing defines from kernel source. Make a diff between file in kernel and backports build to see what is declared in kernel but not on backports..

Too bad that my build tree is now around 90gb so I am not going to do it as I need it to compile something that succeeds instead of failing. It succeeds with this patch and I can assure you that log wouldn't reveal anything surprising; compiler is not doing a thorough investigation, as I, it also knows that it can't use something that it doesn't know about.
If you believe that I made this up, you can just close this.

ynezz commented:
This package is compiled for target, thus the host environment shouldn't matter. It compiles currently fine for everyone, including our buildbots for every platform. So we need to find out, what is different on your side, why it fails to build for you. For the start, the compiler invocation (commandline) with compiler output should be good enough to find out what is going on.

You're not willing to provide that information, I'm not going to spinup Arch container to find out what is going on, so I assume we should stop wasting more time on this. Feel free to ping me again once you've more time/resources working on this.

oskarirauta commented:
@ynezz ping

You are incredibly resilient.
This log is from a month or 2 older git, but results are the same though release of mac80211 is newer on fresh git- I fixed that already back then and when I faced the issue again, with same patch it built nicely again.

```
 ...
  CC [M]  /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/wil6210/wil_platform.o
  CC [M]  /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/wil6210/ethtool.o
  CC [M]  /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/wil6210/wil_crash_dump.o
  CC [M]  /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/wil6210/p2p.o
  LD [M]  /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/wil6210/wil6210.o
  CC [M]  /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/ath10k/mac.o
In file included from /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/backport-include/asm-generic/bug.h:3,
                 from ./arch/x86/include/asm/bug.h:93,
                 from ./include/linux/bug.h:5,
                 from /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/backport-include/linux/bug.h:3,
                 from /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/include/net/mac80211.h:16,
                 from /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/ath10k/mac.h:10,
                 from /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/ath10k/mac.c:8:
/usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/ath10k/mac.c: In function 'ath10k_drain_tx':
/usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/backport-include/linux/lockdep.h:16:47: error: 'LOCK_STATE_HELD' undeclared (first use in this function)
   16 |                         lockdep_is_held(l) == LOCK_STATE_HELD);         \
      |                                               ^~~~~~~~~~~~~~~
./include/asm-generic/bug.h:119:32: note: in definition of macro 'WARN_ON'
  119 |         int __ret_warn_on = !!(condition);                              \
      |                                ^~~~~~~~~
/usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/ath10k/mac.c:4772:9: note: in expansion of macro 'lockdep_assert_not_held'
 4772 |         lockdep_assert_not_held(&ar->conf_mutex);
      |         ^~~~~~~~~~~~~~~~~~~~~~~
/usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/backport-include/linux/lockdep.h:16:47: note: each undeclared identifier is reported only once for each function it appears in
   16 |                         lockdep_is_held(l) == LOCK_STATE_HELD);         \
      |                                               ^~~~~~~~~~~~~~~
./include/asm-generic/bug.h:119:32: note: in definition of macro 'WARN_ON'
  119 |         int __ret_warn_on = !!(condition);                              \
      |                                ^~~~~~~~~
/usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/ath10k/mac.c:4772:9: note: in expansion of macro 'lockdep_assert_not_held'
 4772 |         lockdep_assert_not_held(&ar->conf_mutex);
      |         ^~~~~~~~~~~~~~~~~~~~~~~
make[10]: *** [scripts/Makefile.build:280: /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/ath10k/mac.o] Error 1
make[9]: *** [scripts/Makefile.build:497: /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath/ath10k] Error 2
make[8]: *** [scripts/Makefile.build:497: /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless/ath] Error 2
make[7]: *** [scripts/Makefile.build:497: /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/drivers/net/wireless] Error 2
make[6]: *** [Makefile:1822: /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1] Error 2
make[5]: *** [Makefile.build:13: modules] Error 2
make[4]: *** [Makefile.real:93: modules] Error 2
make[3]: *** [Makefile:121: modules] Error 2
make[3]: Leaving directory '/usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1'
make[2]: *** [Makefile:559: /usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/backports-5.15.8-1/.built] Error 2
make[2]: Leaving directory '/usr/src/openwrt/package/kernel/mac80211'
time: package/kernel/mac80211/regular/compile#300.25#46.51#348.32
    ERROR: package/kernel/mac80211 failed to build (build variant: regular).
make[1]: *** [package/Makefile:116: package/kernel/mac80211/compile] Error 1
make[1]: Leaving directory '/usr/src/openwrt'
make: *** [/usr/src/openwrt/include/toplevel.mk:230: package/mac80211/compile] Error 2
```

Let me know what you find out about that I didn't as it's quite clear to me what is missing.
In my config I have enabled MAC80211_MESH and MAC80211_DEBUGFS, and kmod-mac80211-hwsim

for ath mods I have user_regd, leds, thermal, ct_leds, ct-smallbuffers, smallbuffers, PCOEM as enabled, on everywhere where possible I use openssl instead of alternatives and always support every possible feature available unless it's marked as experimental/testing causing probable problems. Nearly all available modules (excluded lttng, telephony and routing modules) enabled, and on the programs- about 85% is being compiled when telephony and routing is excluded. 
Latest binutils, latest gcc with all the whistles and bells available. Long story short: latest everything if choice is available and still some are even later than that (but that is mostly on package side) - kernel is stock with one addition, I added a patch that suppresses some EFI notifications/errors on early kernel start (during decompression) since they look nasty and mean nothing on virtualised hardware. Some kernel choices that are disabled by default on menuconfig have been enabled, mostly legacy Cgroups and AppArmor stuff which wasn't accepted about a year ago while it still today works flawlessly.

I could provide full .config but it's pretty much easier/faster to read if I just tell crucial parts. Let me know if you want that too... But that's a long long listing.

ynezz commented:
> I could provide full .config but it's pretty much easier/faster to read if I just tell crucial parts.

Output of `./scripts/diffconfig.sh | grep ^CONFIG_KERNEL` should be enough.

oskarirauta commented:
result:
```
subsection "Kconfig recursive dependency limitations"

CONFIG_KERNEL_AUDIT=y
CONFIG_KERNEL_BLK_DEV_BSG=y
CONFIG_KERNEL_BLK_DEV_THROTTLING_LOW=y
CONFIG_KERNEL_BTRFS_FS_POSIX_ACL=y
CONFIG_KERNEL_BUILD_DOMAIN="intrepid"
CONFIG_KERNEL_BUILD_USER="root"
CONFIG_KERNEL_CFQ_GROUP_IOSCHED=y
CONFIG_KERNEL_CGROUP_DEBUG=y
CONFIG_KERNEL_CGROUP_DEVICE=y
CONFIG_KERNEL_CGROUP_FREEZER=y
CONFIG_KERNEL_CGROUP_HUGETLB=y
CONFIG_KERNEL_CGROUP_NET_CLASSID=y
CONFIG_KERNEL_CGROUP_NET_PRIO=y
CONFIG_KERNEL_CGROUP_PERF=y
CONFIG_KERNEL_CIFS_ACL=y
CONFIG_KERNEL_DEBUG_BLK_CGROUP=y
CONFIG_KERNEL_DEFAULT_SECURITY_APPARMOR=y
CONFIG_KERNEL_DEVKMEM=y
CONFIG_KERNEL_DEVMEM=y
CONFIG_KERNEL_DEVTMPFS=y
CONFIG_KERNEL_DEVTMPFS_MOUNT=y
CONFIG_KERNEL_DYNAMIC_DEBUG=y
CONFIG_KERNEL_DYNAMIC_FTRACE=y
CONFIG_KERNEL_ENABLE_DEFAULT_TRACERS=y
CONFIG_KERNEL_EXT4_FS_POSIX_ACL=y
CONFIG_KERNEL_EXT4_FS_SECURITY=y
CONFIG_KERNEL_F2FS_FS_POSIX_ACL=y
CONFIG_KERNEL_F2FS_FS_SECURITY=y
CONFIG_KERNEL_FS_POSIX_ACL=y
CONFIG_KERNEL_FTRACE=y
CONFIG_KERNEL_FTRACE_SYSCALLS=y
CONFIG_KERNEL_FUNCTION_PROFILER=y
CONFIG_KERNEL_FUNCTION_TRACER=y
CONFIG_KERNEL_HFSPLUS_FS_POSIX_ACL=y
CONFIG_KERNEL_HFS_FS_POSIX_ACL=y
CONFIG_KERNEL_HUGETLBFS=y
CONFIG_KERNEL_HUGETLB_PAGE=y
CONFIG_KERNEL_JFFS2_FS_SECURITY=y
CONFIG_KERNEL_KEYS_REQUEST_CACHE=y
CONFIG_KERNEL_KPROBES=y
CONFIG_KERNEL_KPROBE_EVENTS=y
CONFIG_KERNEL_LSM="landlock,lockdown,yama,loadpin,safesetid,integrity,selinux,smack,tomoyo,apparmor,bpf"
CONFIG_KERNEL_MEMCG_SWAP_ENABLED=y
CONFIG_KERNEL_NET_CLS_CGROUP=y
CONFIG_KERNEL_NET_L3_MASTER_DEV=y
CONFIG_KERNEL_NFSD_V2_ACL_SUPPORT=y
CONFIG_KERNEL_NFSD_V3_ACL_SUPPORT=y
CONFIG_KERNEL_NFS_ACL_SUPPORT=y
CONFIG_KERNEL_NFS_V3_ACL_SUPPORT=y
CONFIG_KERNEL_PERF_EVENTS=y
CONFIG_KERNEL_PERSISTENT_KEYRINGS=y
CONFIG_KERNEL_PROC_PAGE_MONITOR=y
CONFIG_KERNEL_PROC_PID_CPUSET=y
CONFIG_KERNEL_REISER_FS_POSIX_ACL=y
CONFIG_KERNEL_RELAY=y
CONFIG_KERNEL_SECURITY=y
CONFIG_KERNEL_SECURITYFS=y
CONFIG_KERNEL_SECURITY_APPARMOR=y
CONFIG_KERNEL_SECURITY_APPARMOR_BOOTPARAM_VALUE=0
CONFIG_KERNEL_SECURITY_NETWORK=y
CONFIG_KERNEL_SECURITY_PATH=y
CONFIG_KERNEL_SLABINFO=y
CONFIG_KERNEL_SLUB_DEBUG=y
CONFIG_KERNEL_SLUB_DEBUG_ON=y
CONFIG_KERNEL_SQUASHFS_XATTR=y
CONFIG_KERNEL_TASKSTATS=y
CONFIG_KERNEL_TASK_DELAY_ACCT=y
CONFIG_KERNEL_TASK_IO_ACCOUNTING=y
CONFIG_KERNEL_TASK_XACCT=y
CONFIG_KERNEL_TMPFS_POSIX_ACL=y
CONFIG_KERNEL_TRANSPARENT_HUGEPAGE=y
CONFIG_KERNEL_TRANSPARENT_HUGEPAGE_ALWAYS=y
CONFIG_KERNEL_UBIFS_FS_SECURITY=y
```

ynezz commented:
Thanks, well I don't see any config option which would in turn enable `CONFIG_LOCKDEP=y` in kernel, anyway I was able to reproduce it and have proposed following [upstream solution](https://lore.kernel.org/backports/20220311194800.452-1-ynezz@true.cz/T/#u).

oskarirauta commented:
Pretty much similar to my patch, except I added 2 more defines to make sure it builds painlessly. So, up to do something for this now on openwrt- or wait for judgement on your proposal?
